### ISSUE TITLE
Datatable: ColumnGroup refactoring

### ISSUE THREAD

**Issue by Rapster** [primefaces/primefaces#6409]
**Description**
Columns grouping follows a very "unconventional" approach: 13 columns have to be defined when only 5 are displayed! My guess, is that no solution have been found at the time to render nested tr/th. Consequence of this, is the current implementation makes the code very hard to maintain at very low level for us as PF developers, but also for developers using column grouping. I can see that ColumnGroup has been giving a really hard time since it's been there 😅 

FYI: I believe other PrimeUI librairies should do the same (if possible)

**Describe the solution you would like**
Could be done 100 times simpler (colspan and rowspan automatically calculated):
```xml
                
<p:dataTable id="tblSales" var="sale" value="#{dtGroupView.sales}" columnGroupLegacyEnabled="false"
                             scrollable="#{dtGroupView.frozenColumns}" frozenColumns="#{dtGroupView.nbFrozenColumns}"
                             scrollWidth="500">

    <f:facet name="header">
        <h:outputText value="Product Sales"/>
    </f:facet>

    <p:column headerText="Product" field="manufacturer" footerText="Table footer (tfoot) using column footer facet"/>

    <p:columnGroup headerText="Sale Rate">
        <p:columnGroup headerText="Sales">
            <p:column headerText="Last Year" field="lastYearSale" footerText="A"/>
            <p:column headerText="This Year" field="thisYearSale" footerText="B"/>
        </p:columnGroup>
        <p:columnGroup headerText="Profit">
            <p:column headerText="Last Year" field="lastYearProfit" footerText="C"/>
            <p:column headerText="This Year" field="thisYearProfit" footerText="D"/>
        </p:columnGroup>
    </p:columnGroup>

    <f:facet name="tfooter">
        <p:row >
            <p:column>Table footer (tfoot) using table tfooter</p:column>
            <p:column colspan="2" style="text-align:right">Totals:</p:column>
            <p:column>
                <h:outputText value="#{dtGroupView.lastYearTotal}">
                    <f:convertNumber type="currency" currencySymbol="$"/>
                </h:outputText>
            </p:column>
            <p:column>
                <h:outputText value="#{dtGroupView.thisYearTotal}">
                    <f:convertNumber type="currency" currencySymbol="$"/>
                </h:outputText>
            </p:column>
        </p:row>
    </f:facet>

    <f:facet name="footer">
        <h:outputText value="Data between 2013-2014"/>
    </f:facet>
</p:dataTable>
```

**Additional context**
Besides it's easier to write and read _(less verbose)_, doing so, No trick is required to make a few features like filtering and sorting work. WDYT? 

**Done**
- Support of frozen columns
- Introducing new facet `tfooter` _(tfoot html tag)_ allowing to define your own tfoot instead of using `ColumnGroup#type`

**Todo**
- [x] Exporters (after #9969)
- [x] Support SubTable
- [x] Support TreeTable
- [ ] Support PanelGrid ?

**To check**
- I think there is a bit of code that will be either dead code, or no longer relevant like:
  -  1dcb97403bdfeaa10055aae7561951f033e56485 introduced getIterableChildren for "perf" reasons but introduced bugs like #2110 which introduced "not so good looking" fixes
  - #2672
  - _To complete if any_

**Comment by melloware** [comment]
I have never used column groups so i will let @mertsincan chime in on this one.

**Comment by melloware** [comment]
If you do refactor we should consider adding ColumnGroup to Exporter: https://github.com/primefaces/primefaces/issues/59

**Comment by Rapster** [comment]
Closing it for now, it seems with recent changes, community is fine with it 😢

**Comment by tandraschko** [comment]
I just fixed bugs for now as RC1 is near
So im fine to talk about refactorings here after the release

**Comment by Rapster** [comment]
ok

**Comment by tandraschko** [comment]
it would be really nice to get rid of ui:repeat support but i think this is a heavy refactoring on how columnGroups works with filters.
This would also make it possible to remove forEachColumn and use FilterMeta#getColumn again

**Comment by tandraschko** [comment]
@Rapster would be great if you can think about it for v11, your are refactorings in DT made it maintainable again!

**Comment by Rapster** [comment]
> it would be really nice to get rid of ui:repeat support but i think this is a heavy refactoring on how columnGroups works with filters.

On the contrary, that's why I didn't want to support it in a first place...
